src/panels/lovelace/components/hui-image.js

```diff
@@ -95,31 +92,49 @@ class HuiImage extends LocalizeMixin(PolymerElement) {
       _imageSrc: null,
       _error: true
     });
+    this.$.image.classList.add('hidden');
   }
 
   _onImageLoad() {
     this._error = false;
+    this.$.image.classList.remove('hidden');
   }
 
   _hassChanged(hass) {
     if (this.cameraImage || !this.entity) {
       return;
     }
 
-    const state = hass.states[this.entity];
-    const unavailable = !isValidObject(state, ['state']);
+    const stateObj = hass.states[this.entity];
+    const newState = !stateObj ? 'unavailable' : stateObj.state;
+
+    if (newState === this._currentState) return;
+    this._currentState = newState;
 
+    this._updateStateImage();
+    this._updateStateFilter(stateObj);
+  }
+
+  _updateStateImage() {
     if (!this.stateImage) {
-      this._imageClass = unavailable || STATES_OFF.includes(state.state) ? 'state-off' : '';
+      this._imageFallback = true;
       return;
     }
+    const stateImg = this.stateImage[this._currentState];
+    this._imageSrc = stateImg || this.image;
+    this._imageFallback = !stateImg;
+  }
 
-    const stateImg = !unavailable ? this.stateImage[state.state] : this.stateImage.unavailable;
+  _updateStateFilter(stateObj) {
+    let filter;
+    if (!this.stateFilter) {
+      filter = this.filter;
+    } else {
+      filter = this.stateFilter[this._currentState] || this.filter;
+    }
 
-    this.setProperties({
-      _imageClass: !stateImg && (unavailable || STATES_OFF.includes(state.state)) ? 'state-off' : '',
-      _imageSrc: stateImg || this.image
-    });
+    const isOff = !stateObj || STATES_OFF.includes(stateObj.state);
+    this.$.image.style.filter = filter || (isOff && this._imageFallback && DEFAULT_FILTER) || '';
```

Comment:
This shit is getting so crazy, I feel like we need to add tests to make sure that all use cases are getting covered.